### ISSUE TITLE
Implement efficient foward derivatives and generalize DualNumber.hs definitions to work on them

### ISSUE THREAD

**Issue by Mikolaj** [Mikolaj/horde-ad#15]
[Edit: current status:

In short the randomized testing remains, as described below, using @awf 's formulas.

Long version: Thanks to advice from the last meeting, everything is now overhauled and the main class now takes the dual component type as argument (e.g., Delta0 Double), not the primal component type (e.g., Double). This is much better than the newtype approach, though is has many drawbacks of its own:

https://github.com/Mikolaj/horde-ad/blob/d069a45773ed849913b5ebd0345153072f304fd9/src/HordeAd/Core/DualClass.hs#L69-L101

]

Original description:

Then write QuickCheck tests that compute forward derivatives in the efficient way and using `evalBindingsForward` and assert that they are within a tiny floating point fuzz. If easy, also compute gradients and try to relate them via jacobians (one gradient run per base vector? or one forward derivative? sounds expensive either way).

This is easy for mathematical operations (same tensor rank, only linear structure used in delta expressions), but hard for operations spanning ranks.

[Edit2: an icomplete catalogue of approaches that I tried and that failed:
* `data DualNumber a = D (Primal a) (Dual a)`: fails as described at https://github.com/Mikolaj/horde-ad/issues/15#issuecomment-1066826756
* `class HasRanks (forwardDirection :: Bool) r where`: fails, because the associated type families ignore `forwardDirection` and require the same result
* `unsafeCoerce` of arguments to `dfastForwardShow`: fails, because while it fixes `dfastForwardShow`, the very `DualNumber` functions still require Numeric instance to compute their primal component
* the desperate conflicting instances in separate modules: fail, because type family clashes are checked globally, even when imports don't clash
* newtype of Double works, but has various problems (see "Old status after abandoning the experiment and using newtypes instead" below), so is abandoned]

**Comment by awf** [comment]
One can also use the relation that if
```
f : S -> T
f_fwd : (S, dS) -> dT
f_bwd : (S, dT) -> dS
```
then there is a consistency relation for `s : S, ds : dS, dt : dT`
```
dot(f_fwd(s, ds), dt) = dot(f_bwd(s, dt), ds)
```
If that holds for random `ds`, `dt`, there's a good chance the derivatives are correct.

**Comment by Mikolaj** [comment]
Oh, indeed, we don't need to check for all vectors from a base, which may be millions, but only for a number of completely random vectors. So we QuickCheck not only over random inputs of the objective function, but also over random infinitesimals of both kinds. This was supposed to be stochastic testing anyway.

**Comment by awf** [comment]
Indeed, and they don't even need to be "small" - those relationships are true even if ds, dt have unit norm.

**Comment by awf** [comment]
But ideally draw them from e.g. a Gaussian, in case unit norm hides a scaling bug

**Comment by Mikolaj** [comment]
If we managed to pull this off (which is probably doable by forking hmatrix, but maintaining that fork would be a huge burden), we should also be able to define special dual number instances for computing exclusively the value of the objective function (#11), getting it very close to the speed of a normal Haskell numerical function (currently we still have lots of overhead, despite being order of magnitude faster than when computing gradients).

**Comment by Mikolaj** [comment]
> But ideally draw them from e.g. a Gaussian, in case unit norm hides a scaling bug

Is uniform distribution acceptable? Or do we want to focus on near-zero values? I don't know how hard it is to get Gaussian or normal distributions for QuickCheck (it's probably trivial, one just applies an exponent and multiplies by pi, but my maths speciality was logic, not calculus, so I wouldn't know).

**Comment by awf** [comment]
Uniform -5..5 is fine.

**Comment by Mikolaj** [comment]
Thank you. I've eventually settled on uniform distribution with range (-r, r), where r is random (independently for parameters and ds) between 0.01 and 10. Anyway, there is a lot of randomness in the MNIST fcnn test for which I implemented the QuickCheck comparison of two methods of forward derivatives and the gradient: both hidden layers have random sizes, glyph and label are random (not from real data), parameters and ds are random, both generated using a uniform distribution with random variances as above, 

BTW, the test shows that the MNIST fcnn variant with fused SoftMax and CrossEntropy gives different derivatives (but equal objective function values) as the variant with separately applied SoftMax and CrossEntropy. That was no detected by long gradient descent runs with realistic parameters (and both real and random/fake MNIST data). I'm guessing this may be the overflow of exponentiation you warned about. If so, I will finally have a test that verifies and protects the ` expU = exp (u - HM.scalar (V.maximum u))` adjustment and so I can add it with confidence. TODO.
